### PR TITLE
experiment with widget extension

### PR DESCRIPTION
**Pull request by nenb**
## Reference Issues or PRs

https://github.com/nebari-dev/jupyterlab-jhub-apps/issues/4

## What does this implement/fix?

_Put a `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds a feature)
- [x] Breaking change (fix or feature that would cause existing features not to work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

## Testing

- [ ] Did you test the pull request locally?
- [ ] Did you add new tests?

#### Text styling

- [ ] The content is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [ ] If there are headers, they use the proper header tags (with only one level-one header: `H1` or `#` in markdown).
- [ ] All links describe where they link to (for example, check the [Nebari website](https://nebari.dev/)).
- [ ] This content adheres to the Nebari style guides.

#### Non-text content

- [ ] All content is represented as text (for example, images need alt text, and videos need captions or descriptive transcripts).
- [ ] If there are emojis, there are not more than three in a row.
- [ ] Don't use [flashing GIFs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [ ] If the content were to be read as plain text, it still makes sense, and no information is missing.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -17,14 +16,55 @@ interface IDeployAppArgs {
...
-    const openURL = (url: string) => {
+    const deploymentFormUrl = '/services/japps/create-app';
+
+    const openInIFrame = (deployUrl: string) => {
+      try {
+        const iframe = document.createElement('iframe');
+        iframe.src = deployUrl;
+        iframe.width = '100%';
+        iframe.height = '100%';
+        iframe.style.border = 'none';
```
Comment (krassowski) [inline src/index.ts:56]: JupyterLab includes [`IFrame`](https://jupyterlab.readthedocs.io/en/stable/api/classes/apputils.IFrame-1.html?highlight=Iframe#) class; you do not have to use it but you it includes good security defaults (`sandbox` - some of the options may need disabling) and sets the width/height to 100% by default. In the next version if will also expose `lazy` `loading` attribute which is nice as it makes JupyterLab load faster when multiple tabs with iframes are open.

File: src/index.ts [hunk 2]
```diff
@@ -17,14 +16,55 @@ interface IDeployAppArgs {
...
+
+        const widget = new deployAppWidget();
+        widget.node.appendChild(iframe);
```
Comment (krassowski) [inline src/index.ts:59]: If creating a custom widget subclass I would create iframe inside of the widget constructor; I would pass all options that are public API (say the deploy URL) via an `IOptions` interface object (objects are easy to extend and individual properties can be deprecated/removed limiting public API changes)

File: src/index.ts [hunk 3]
```diff
@@ -17,14 +16,55 @@ interface IDeployAppArgs {
...
+
+        app.shell.add(widget, 'main');
```
Comment (krassowski) [inline src/index.ts:61]: You could (but don't have to) wrap your widget in [`MainAreaWidget`](https://jupyterlab.readthedocs.io/en/stable/api/classes/apputils.MainAreaWidget-1.html?highlight=MainAreaWidget); it mostly adds a spinner for when the inner widget is loading and integrates it with the printing utility.

File: src/index.ts [hunk 4]
```diff
@@ -67,3 +118,13 @@ const jhubAppsPlugin: JupyterFrontEndPlugin<void> = {
...
 
 export default plugins;
+
+class deployAppWidget extends Widget {
+  constructor() {
+    super();
+    this.addClass('jp-example-view');
+    this.id = 'deploy-app';
+    this.title.label = 'Deploy App';
+    this.title.closable = true;
```
Comment (krassowski) [inline src/index.ts:128]: I think we would want to have an icon in the tab bar, we can reuse the same rocket icon for now.

**Review by krassowski — COMMENTED** [review]
Exciting :rocket: 

To follow the best lab practices we will want to use `ILayoutRestorer` to keep the widget if the browser refreshes/looses connection see https://jupyterlab.readthedocs.io/en/stable/extension/extension_tutorial.html#restore-panel-state-when-the-browser-refreshes - this would be in scope for this PR.

Ideally we would also keep the state of the form inside the iframe itself. This is out of scope for this PR, I will open an enhancement issue

**Comment by nenb** [comment]
Closing in favour of #12

**Pull request closed without merging**